### PR TITLE
feat: add Keycloak idp support

### PR DESCRIPTION
**Pull request by seriouszyx**
Signed-off-by: Yixiang Zhao <seriouszyx@foxmail.com>

### CODE REVIEW

**Comment by hsluoyz** [comment]
@Steve0x2a plz review

**Comment by hsluoyz** [comment]
@seriouszyx demo link?

**Comment by seriouszyx** [comment]
@hsluoyz 
- Casdoor demo link: http://82.156.216.21:8000/login 
  - Aliyun
    - username: admin
    - password: Admin123456!
  - Keycloak
    - username: admin
    - password: 123
- Keycloak demo link: http://82.156.216.21:8080/auth  Administrator console
  - username: admin
  - password: 123
  - realm: Master
  - clients name: `http://82.156.216.21:8000/api/acs `

**Comment by hsluoyz** [comment]
@seriouszyx the demo looks great! Just a minor issue in profile UI: http://82.156.216.21:8000/account

image

File: object/saml.go [hunk 1]
```diff
@@ -97,3 +90,21 @@ func buildSp(provider *Provider, samlResponse string) (*saml2.SAMLServiceProvide
...
+	tagMap := map[string]string{
+		"Aliyun IDaaS": "ds",
+		"Keycloak": "dsig",
+	}
+	tag := tagMap[providerType]
+	expression := fmt.Sprintf("<%s:X509Certificate>([\\s\\S]*?)</%s:X509Certificate>", tag, tag)
+	res := regexp.MustCompile(expression).FindStringSubmatch(deStr)
+	str := res[0]
+	tagLength := len("<:X509Certificate>") + len(tag)
+	return str[tagLength : len(str) - tagLength - 1]
```
Comment (Steve0x2a) [inline object/saml.go:109]: `return res[1]` can get the same result and avoid using slices.

**Review by Steve0x2a — CHANGES_REQUESTED** [review]
lgtm

**Comment by seriouszyx** [comment]
@hsluoyz Fixed.

image

**Comment by hsluoyz** [comment]
lgtm

**Pull request merged**